### ISSUE TITLE
Should we move to GitLab?

### ISSUE THREAD

**Issue by sduenas** [chaoss/grimoirelab#296]
As part of the discussion in #284 (284#issuecomment-588211843) I suggested the possibility of moving GrimoireLab (and probably CHAOSS) to GitLab.

As we all know, GitHub is a great platform but it's not open source. In GrimoireLab we are committed to FLOSS philosophy. Everything we produce and we use is free/libre/open source software. Everything but GitHub. GitHub is not open source.

When we started with the former project (MetricsGrimoire) there wasn't any other viable alternative to GitHub so we decided to use it. During the last few years GitLab has evolved a lot, providing a strong platform full of features. In Bitergia, we have been using it for 2 years and we're really happy about it.

So, my question is: should we move to GitLab? My main reasons about it are:

1) Use a partially open source platform. I understand we are not going to have a self-hosted server, so we will use gitlab.com service. I guess not everything there is open source but mostly of it will be.
2) Strong community using and working on it.
3) A lot of tools integrated with the platform (CI/CD, Kanban, etc)
3) We can organize the project in groups and sub-projects fixing problems like the one we're discussing in #284 

The main problems that I see are:

1) Travis.CI is not supported. So far, we have been using Travis to run automated tests to verify if the tools work as expected. Travis is not supported in GitLab because they use their own system. Besides updating configurations, the main problem there are quotas and limit usage time of GitLab's CI so we might end paying for their services.
2) Code review works different so we'll need some time to adjust to the new interface

Having said that, I'd like to open this question to everybody and here comments about whether we should move or not? At the same time we can open the same discussion in CHAOSS project.

**Comment by alpgarcia** [comment]
My questions are: 
* Could we move GrimoireLab to GitLab even if other CHAOSS projects are still on GitHub or we need to make a joint decision?
* Could we have CHAOSS organization on both platforms so each project can decide its own way?

**Comment by sduenas** [comment]
> My questions are:
> 
>     * Could we move GrimoireLab to GitLab even if other CHAOSS projects are still on GitHub or we need to make a joint decision?
> 

We can communicate our decision to the board and I think they will need to decide what to do.

>     * Could we have CHAOSS organization on both platforms so each project can decide its own way?

Probably but I think the board will need to decide. I still think it was a bad idea to move GrimoireLab repositories under CHAOSS organization when we already had one. I understand the reasons, though. In GitLab, our project can be organized in a subgroup or subproject among the other repositories.

**Comment by lukaszgryglicki** [comment]
I prefer GitHub, but it's not me to decide.

**Comment by geekygirldawn** [comment]
It's probably a good idea to have some of this discussion on the CHAOSS mailing list, and maybe in the Tuesday CHAOSS project meeting?

**Comment by sduenas** [comment]
I won't be available next Tuesday but maybe you can talk about it. Btw, I've just sent a message the the mailing list, so we can discuss about it there too.

**Comment by GeorgLink** [comment]
Thanks for bringing the discussion to the mailing list.

**Comment by penyuan** [comment]
I'm a very new user who is starting to learn about, make use of, and ideally contribute in a small way to the GrimoireLab repositories.

First of all, in general I think it is a good idea to move from a completely closed source platform (Github) to one that is slightly more open source friendly and at least partly open source (Gitlab). This is indeed in line with the values behind CHAOSS.

Just to add a data point to whether to move to Gitlab, to my knowledge Github is blocked in several countries. By moving to Gitlab (or ideally to a decentralised self-hosted Gitlab (or other git) instance) could make CHAOSS/Grimoirelab development community more inclusive.

**Comment by GeorgLink** [comment]
This is interesting. Do you have a news article or something that describes which countries GitHub is blocked in?

**Comment by penyuan** [comment]
I remember the Github blocking problem raised in an issue for another active Github repository, unfortunately I can't remember which repository! However, I clearly recall that it was an Iranian developer who really want to contribute but couldn't because Github is blocked in Iran (at least at the time, i.e. last year).

I quickly searched for some examples, though:

* [GitHub starts blocking developers in countries facing US trade sanctions](https://www.zdnet.com/article/github-starts-blocking-developers-in-countries-facing-us-trade-sanctions/)
* [Censorship of Github](https://en.wikipedia.org/wiki/Censorship_of_GitHub) on Wikipedia
* [GitHub confirms it has blocked developers in Iran, Syria and Crimea](https://techcrunch.com/2019/07/29/github-ban-sanctioned-countries/)

In any case, I think it's at least worth considering.

**Comment by jsmanrique** [comment]
+1 to move to GitLab. The capability of using groups and subgroups, [issue boards](https://docs.gitlab.com/ee/user/project/issue_board.html) where all the issues of a group and subgroups can be managed together or by subgroup, the amazing [Web IDE](https://docs.gitlab.com/ee/user/project/web_ide/) that potentially eases the contribution process al lot, the awesome [markdown support](https://docs.gitlab.com/ee/user/markdown.html), including the capability of drawing [diagrams & workflows](https://docs.gitlab.com/ee/user/markdown.html#diagrams-and-flowcharts) using just *text* ... Those are features I love the most.

**Comment by dicortazar** [comment]
Quick comment here so everyone in the medium-term future is aware: this ticket has been discussed here, as threads in the mailing list, and during the weekly CHAOSS meetings.

The final decision was to move at least GrimoireLab there and check frictions we may face.

**Comment by GeorgLink** [comment]
A final decision in the CHAOSS community still needs to be reached, but the chances are very high that we will be moving to GitLab.

**Comment by sduenas** [comment]
I made some research about how to migrate to GitLab. It will be pretty straight forward because there's an option to do that when a new repository is created. The only thing we need to is to allow GitLab to get access to the `chaoss` group.

The part that will be more difficult is to configure GitLab CI. If we migrate to GitLab we won't be able to use Travis CI (at least as we do it now), so we should use GitLab CI instead to avoid integration problems. It took me about 1,5 to read some documentation and to make my jobs work. I use as a test the `muggle` branch of SortingHat. This branch contains code in python and javascript, plus use a MySQL service. This is the `.gitlab-ci.yml` file that I needed to create:

```
stages:
  - test

python-test:
  stage: test
  only:
    refs:
      - muggle
  image: python:3.6
  services:
    - mysql:5.7
  variables:
    MYSQL_DATABASE: "root"
    MYSQL_ALLOW_EMPTY_PASSWORD: "true"
  before_script:
    - pip install flake8
  script:
    - pip install -r requirements.txt
    - python manage.py test --settings=config.settings.testing

ui-test:
  stage: test
  only:
    refs:
      - muggle
  image: node:10.20
  cache:
    paths:
      - ui/node_modules/
      - ui/.yarn
  script:
    - cd ui
    - yarn install
    - yarn test:unit
```

For this case I had to modify the configuration of the django settings to set the mysql host name as `mysql` instead of `localhost`. In GitLab CI services endpoints use the same name of the service.
I also made some research about how to configure `coveralls` but I didn't try it. This is what we will need:
- https://docs.coveralls.io/gitlab

Basically, it consists on calling coveralls: https://gitlab.com/nickmerwin/coveralls-demo/-/blob/master/Makefile#L27

**Comment by sduenas** [comment]
After the discussion with our CHAOSS folks during the weekly meeting we decided we're going to still use GitHub for at least six months. After that time we can reconsider it again.

The rationale behind this is the cost of moving to a different solution and the problems that it may cause to the CHAOSS working groups. Other members don't have strong reasons to move. They are happy to stay on GitHub.

So from here, we need to decide how to fix all the problems that we have using GitHub, specially #284.
